openlibrary/components/BulkSearch/utils/classes.js

```diff
@@ -122,6 +123,49 @@ export class AiExtractor extends AbstractExtractor{
     }
 }
 
+export class TableExtractor extends AbstractExtractor{
+
+    name = 'table_extractor'
+    /**
+     *
+     * @param {string} label
+     */
+    constructor(label) {
+        super(label)
+        /** @type {string} */
+        this.authorColumn = 'author'
+        /** @type {string} */
+        this.titleColumn = 'title'
+    }
+
+    /**
+     * @param {ExtractionOptions} extractionOptions
+     * @param {string} text
+     * @return {Promise<BookMatch[]>}
+     */
+    async run(extractionOptions, text){
+
+        /** @type {string[]} */
+        const lines = text.split('\n')
+        /** @type {string[][]} */
+        const cells = lines.map(line => line.split('\t'))
+        /** @type {{columns: String[], rows: {columnName: string}[]}} */
+        const tableData = {
+            columns: cells[0],
+            rows: []
+        }
+        for (let i=1; i< lines.length; i++){
+            const row = {}
+            for (let j = 0; j < tableData.columns.length; j++){
+                row[tableData.columns[j].trim().toLowerCase()] = cells[i][j]
+            }
+            // @ts-ignore
+            tableData.rows.push(row)
+        }
+        return tableData.rows.map(row => new BookMatch(new ExtractedBook(row?.[this.authorColumn] || '', row?.[this.titleColumn] || ''), {}))
```

Comment:
row itself will never be null so likely don't need these. Also: some newlines here would be helpful for readability!

```suggestion
        return tableData.rows.map(row => new BookMatch(new ExtractedBook(row[this.authorColumn] || '', row[this.titleColumn] || ''), {}))
```

Reply:
Resolved